tests/trogdor/produce_spec_runner.py

```diff
@@ -0,0 +1,222 @@
+# "targetMessagePerSec": #Java int#
+# "maxMessages": #Java long#
+# "keyGenerator": #payloadGenerator, optional, sequential generator (default)#
+#  { "type": "constant" | "sequential" | "uniformRandom" | "null" }
+# "valueGenerator"#payloadGenerator, optional, constant generator (default)#"
+#  {"type": "constant" | "sequential" | "uniformRandom" | "null" }
+# "transactionGenerator":#Object, optional, empty (default)# {"type":"uniform"}
+# "producerConf": #a JSON object#
+# "commonClientConf": "a JSON object"
+# "adminClientConf": "a JSON object"
+# "activeTopics" : "a JSON object"
+# "inactiveTopics": "a JSON object"
+#
+# PayloadGenerator:
+# ConstantPayLoadGenerator:
+# "type": #string, "constant"
+# "size": #Java Int, "size of the value"
+# "value": #Java byte array, optional, "constant value"
+# SequentialPayloadGenerator:
+# "type": "sequential"
+# "size": #Java Int, "size of the value"
+# "startOffset": #Java Long, "start value of the increasing value"
+# uniformRandomPayloadGenerator:
+# "type": "uniformRandom"
+# "size": #Java Int", "size of the value"
+# "seed": #Java Long", "seed of the value"
+# "padding" #Java Long", #padding size of the value, size - padding is the size of random bytes"
+import math
+import threading
+import time
+
+from hdrh.histogram import HdrHistogram
+
+from trogdor_utils import expand_topics, trogdor_log, merge_topics, create_topics, \
+    update_trogdor_status, create_admin_conn, partition_set, PayloadGenerator, create_producer_conn, \
+    get_payload_generator
+
+
+def execute_produce_spec(workload):
+    """ Return after finishing the workload """
+    update_trogdor_status("running")
+    runner = ProduceSpecRunner(workload)
+    runner.monitor()
+
+
+class ProduceSpecRunner:
+    def report_status(self, realQPS=None):
+        """ Report Histogram Latency"""
+        exp_status = {"totalSent": self.nr_finished_messages,
+                      "totalRecorded": self.latency_histogram.get_total_count(),
+                      "totalError": self.nr_failed_messages,
+                      "planMPS": self.qps,
+                      "averageLatencyMs": self.latency_histogram.get_mean_value()/100.0,
+                      "p50LatencyMs": self.latency_histogram.get_value_at_percentile(50)/100.0,
+                      "p95LatencyMs": self.latency_histogram.get_value_at_percentile(95)/100.0,
+                      "p99LatencyMs": self.latency_histogram.get_value_at_percentile(99)/100.0,
+                      "maxLatencyMs": self.latency_histogram.get_max_value()/100.0
+                      }
+        if realQPS:
+            exp_status["realMPS"] = realQPS
+        update_trogdor_status(exp_status)
+
+    def message_on_delivery(self, err, msg, sent_time):
+        if err is not None:
+            trogdor_log("ProduceSpecRunner: delivery failed: {} [{}]: {}".format(msg.topic(), msg.partition(), err))
+            self.nr_failed_messages += 1
+        now = time.time()
+        latency = now - sent_time
```

Comment:
I am working on adding an `iteration` field to the produce spec. With `iteration`, you can control how many times Trogdor should run the same workload. That's a better way to compare `warmup` Vs. `stable` cases.

Reply:
That is probably true for the cluster itself, but each client instance will need some warmup to amortize for connection setup times, etc, and this should not be included in the final throughput or latency results.

Reply:
The warmup will be the first iteration, say first 5000 messages. Also, the warmup latency is important, I suggest we should show that too.

Reply:
Are the iterations using the same producer instance?
If so, then this is fine, but if new producer instances are used for each iteration we still need a warmup period per iteration to avoid including startup costs in the throughput measurements.

Reply:
Any comment on this?

Reply:
I added the `iterations` field to the spec file.  When the field > 1, the trogdor_runner.py reports an JSON array of performance results, e.g. {"status": [{"totalSent": 50000, "totalRecorded": 50000, "totalError": 0, "planMPS": 10000, "averageLatencyMs": 13814.083504000002, "p50LatencyMs": 13864.95, "p95LatencyMs": 29470.71, "p99LatencyMs": 30719.99, "maxLatencyMs": 31006.71}, {"totalSent": 50000, "totalRecorded": 50000, "totalError": 0, "planMPS": 10000, "averageLatencyMs": 19482.478866999998, "p50LatencyMs": 19394.55, "p95LatencyMs": 35573.75, "p99LatencyMs": 36597.75, "maxLatencyMs": 36823.03}]}

Reply:
Are the iterations reusing the same producer instance?
If not then there is no warm-up benefit on the client side, so I really think they should be reused.

Reply:
Yeah, they use the same producer.